### PR TITLE
90: Implement top level blocksRange queries.

### PR DESCRIPTION
**Pull request by akhila-raju**

### CODE REVIEW

File: src/resolvers/block.ts [hunk 1]
```diff
@@ -1,6 +1,8 @@
 import { IFieldResolver, IResolvers } from 'graphql-tools';
 import * as _ from 'lodash';
 import { web3 } from '../providers/web3';
+import * as util from 'util';
```
Comment (raulk) [inline src/resolvers/block.ts]: Can we get rid of this import?

File: src/resolvers/block.ts [hunk 2]
```diff
@@ -36,22 +38,66 @@ const blocks: IFieldResolver<any, any> = (obj, { numbers, hashes }: IBlocksArgs)
...
+interface IBlocksRangeArgs {
+  numbers?: number[];
+  hashes?: string[];
+}
+
+const blocksRange: IFieldResolver<any, any> = async (obj, { numbers, hashes }: IBlocksRangeArgs) => {
+  if (numbers && hashes) {
+    throw new Error('Only one of blocks or hashes should be provided.');
+  }
+  type funcType = (hash: string, callback: (err: Error, res: BlockWithoutTransactionData) => void) => void;
```
Comment (raulk) [inline src/resolvers/block.ts]: I think this type declaration can go.

File: src/resolvers/block.ts [hunk 3]
```diff
@@ -36,22 +38,66 @@ const blocks: IFieldResolver<any, any> = (obj, { numbers, hashes }: IBlocksArgs)
...
+  
+  const getBlock: ((string) => Promise<BlockWithoutTransactionData>) = (hash) => {
+    return new Promise((res, rej) => {
+      web3.eth.getBlock(hash, (err, block) => {
+        if (err) {
```
Comment (raulk) [inline src/resolvers/block.ts]: I think we can turn this if-else block into a one-liner: `err ? rej(err) : res(block)`

File: src/resolvers/block.ts [hunk 4]
```diff
@@ -36,22 +38,66 @@ const blocks: IFieldResolver<any, any> = (obj, { numbers, hashes }: IBlocksArgs)
...
+      });
+    });
+  };
+
+  let start: number, end:number;
+
+  if (numbers && numbers.length == 2) {
+    [start, end] = numbers;
+  } else if (hashes && hashes.length == 2) {
+    const startBlock = await getBlock(hashes[0]);
```
Comment (raulk) [inline src/resolvers/block.ts]: Could we fetch both blocks concurrently? For example:

```
[start, end] = await Promise.all(hashes.map(getBlock)).map(b => b.number);
```

File: src/schema/core.graphql [hunk 5]
```diff
@@ -12,7 +12,7 @@ type Query {
   blocks(numbers: [Int], hashes: [String]): [Block]
 
   "Selects a range of blocks."
-  blocksRange(numberRange: [Int], hashRange: [String]): [Block]
+  blocksRange(numbers: [Int], hashes: [String]): [Block]
```
Comment (raulk) [inline src/schema/core.graphql]: Try not to change the schema. The schema is the contract with our clients, and we don't want to change it as the result of implementation details. Can we restore the previous field names?

File: src/resolvers/block.ts [hunk 6]
```diff
@@ -36,22 +38,66 @@ const blocks: IFieldResolver<any, any> = (obj, { numbers, hashes }: IBlocksArgs)
...
     blocks,
+    blocksRange,
   },
   Block: {
     transactions,
   },
 };
 
-export default resolvers;
+export default resolvers;
```
Comment (raulk) [inline src/resolvers/block.ts]: Are these line breaks being added by some tool? Are they added by the linter on commit? Or by your formatting configuration in VS Code? These changes muddy up the diff a little bit, so it's a good idea to review your diff before making the commit, and reverting such changes if possible.

File: src/__tests__/core.test.ts [hunk 7]
```diff
@@ -72,3 +72,34 @@ it('Test multiple blocks selected by specific hashes', async () => {
...
+      }
+    }
+  `;
+
+  const expected = {
+    data: { blocksRange: [{ number: 20222 },{ number: 20223 },{ number: 20224 }] } };
+
+  const result = await graphql(schema, query);
+  expect(result).toEqual(expected);
+});
```
Comment (raulk) [inline src/__tests__/core.test.ts]: I'm missing negative tests here. For example:

* Provide both number and hash ranges => error.
* Provide more than two numbers in the range => error.
* Provide more than two hashes in the range => error.
* Provide no numbers or hashes => error.
* Provide 1 number only => error.
* Provide 1 hash only => error.
* Provide a negative number in the range => error.
* Provide a start number larger than the end number => error.
* Provide a start hash whose block number is higher than the end number => error.
* Provide a hash that does not exist => error.
* Provide the same block number twice => it should just return one block.
* Provide the same hash twice => it should just return one block.

File: src/resolvers/block.ts (hunk 1 above)
Comment (akhila-raju) [reply src/resolvers/block.ts]: Resolved

File: src/resolvers/block.ts (hunk 4 above)
Comment (akhila-raju) [reply src/resolvers/block.ts]: This errors out - "Promise.all(...).map is not a function"

File: src/resolvers/block.ts (hunk 3 above)
Comment (akhila-raju) [reply src/resolvers/block.ts]: Resolved

File: src/resolvers/block.ts (hunk 2 above)
Comment (akhila-raju) [reply src/resolvers/block.ts]: Resolved

File: src/resolvers/block.ts (hunk 6 above)
Comment (akhila-raju) [reply src/resolvers/block.ts]: Strange. Not sure what's causing it but will double check in the future.

**Comment by raulk** [comment]
[ch90]

**Pull request merged**